scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+        """
+        query_job = self.bigquery_client.query(query)
+        return {row[0] for row in query_job}
+
+    # local csv interactions
+    def create_csv(self, row_ids: Set[str], target_table: str) -> str:
+        """
+        Create a csv locally with one column filled with row ids to soft delete.
+        :param row_ids: A set of row ids to soft delete.
+        :param target_table: The table that the row ids belong to.
+        :return: The filename of the created csv.
+        """
+        filename = self.filename_template.format(table=target_table)
+        with open(filename, mode="w") as soft_delete_file:
+            sd_writer = csv.writer(soft_delete_file, delimiter=",", quotechar='"', quoting=csv.QUOTE_MINIMAL)
+
+            for rid in row_ids:
+                sd_writer.writerow([rid])
+
+        return filename
+
+    @staticmethod
+    def delete_csv(filename: str):
+        """
+        Remove the csv after it has been uploaded so that there isn't a deluge of csv files in the local directory.
+        :param filename: The filename of the file to delete.
+        :return:
+        """
+        os.remove(filename)
+
+    # gcs (cloud storage) interactions
+    def put_csv_in_bucket(self, filename: str) -> str:
+        """
+        Puts a local file into a GCS bucket accessible by Jade.
+        :param filename: The filename of the file to upload.
+        :return: The gs-path of the uploaded file.
+        """
+        storage_client = storage.Client(project=self.bucket_project, credentials=self.gcp_creds)
+        bucket = storage_client.bucket(self.bucket)
+        blob = bucket.blob(filename)
+        blob.upload_from_filename(filename)
+
+        filepath = f"gs://{self.bucket}/{filename}"
+        print(f"Put a soft-delete file here: {filepath}")
+
+        return filepath
+
+    # jade interactions
+    @staticmethod
+    @lru_cache()
+    def _get_endpoint(endpoint: str):
+        """
+        Get relevant information, such as the HTTP method, the path, and parameters, for Jade API endpoints.
+        :param endpoint: The endpoint name to get information for.
+        :return: A dictionary containing the endpoint's method, path, and parameters.
+        """
+        url = "https://jade-terra.datarepo-prod.broadinstitute.org/v2/api-docs"
+        response = CachedSession().get(url=url)
+        paths = response.json()["paths"]
+
+        my_map = {}
+
+        for path, methods in paths.items():
+            for method, rest in methods.items():
+                if "parameters" in rest:
+                    params = [(param["name"], param["in"]) for param in rest["parameters"]]
+                else:
+                    params = []
+                my_map.update({rest["operationId"]: {"method": method.upper(), "path": path, "params": params}})
+        if endpoint in my_map:
+            return my_map[endpoint]
+        else:
+            raise KeyError(f"Endpoint named {endpoint} not found!")
```

Comment:
```suggestion
        try:
            return my_map[endpoint]
        except KeyError:
            raise KeyError(f"Endpoint named {endpoint} not found!")
```

weirdly, python best practice is to ask forgiveness, not permission

Reply:
+1, A bit more color can be found in the python [docs](https://docs.python.org/3/glossary.html#term-eafp)